DTable implementation

krynju opened the pull request:
TODO:

- [x] Tables.jl compatible constructors 
- [x] Constructor for prepartitioned data (proto for now, needs the planned storage api for full implementation)
- [ ] fetch(d::DTable) into any Tables.jl data structure (currently back to DataFrame)
- [ ] Implement the Tables.jl interface for the DTable (+ figure out what makes sense to implement and what doesn't)
- [ ] add more TODO items

jpsamaroo commented on Project.toml:
```diff
@@ -4,6 +4,7 @@ version = "0.11.3"
 
 [deps]
 Colors = "5ae59095-9a9b-59fe-a467-6f913c188581"
+DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
```
We should put the DataFrames-specific code behind an `@requires` block in `__init__()` so that we don't take on a DataFrames dependency. We should only need the Tables direct dependency.

krynju replied:
I've put it in the @requires block, but then the exported functions and structures are not available after importing the package into env.  I'll check out later how to handle this

jpsamaroo reviewed (changes requested):
Great work so far!

jpsamaroo commented:
@krynju when you get the chance, let's try benchmarking this against raw DataFrames and see how we do. I suspect we'll probably want some scheduler changes to optimize concurrent thunk submission (which currently takes a global scheduler lock).

ExpandingMan commented:
Hey guys, glad to see this exciting work, thanks for all the effort!

I feel a little awkward making such a radical suggestion as merely an interested observer, but are you sure it would not be better to split this into a new package dependent on Dagger, rather than relying on `@requires`?  This approach has worked out quite well in the JuliaData ecosystem where many packages have delightfully minimal dependencies: for example several extensions to DataFrames.jl exist outside that package such as DataFramesMeta.jl, IO packages such as CSV.jl which despite being used with DataFrames.jl in the majority of use cases only rely on the Tables.jl interface.  There are even alternative tables packages such as TypedTables.jl which implement the Tables interface and are completely independent in terms of dependencies.  If I make any significant contribution to the distributed tables stuff myself, it will likely be in expanding the functionality of Parquet.jl to be compatible with your work here, and it's not yet clear to me whether this will involve dependencies on DTables at all, perhaps it will only need Dagger and perhaps it will only need Tables.

I realize that currently this isn't a lot of code, so splitting it may seem premature, but it could easily grow over time to be an enormous code base, so splitting it can make dependency logic much more straightforward, and it is often easier to split projects in their nascent stages rather than having a massive reorganization later on.

krynju commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,46 @@
+
+mutable struct DTable
+    v::Vector{Dagger.EagerThunk}
+    thunksize::Int
```
I agree, I've added this at some point and didn't really think about it.
The chunk/partition size would only be relevant on creation and only in case the source needs to be partitioned by the constructor.
At this point I'll remove it as I don't really see a point in keeping it either

krynju commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,46 @@
+    thunksize::Int
+end
+
+function DTable(table::DataFrames.DataFrame; thunksize=10)
```
Yes, Tables.jl input and making the DTable Tables.jl compatible is definitely on the todo list.
Tried to do it from the start, but between DataFrames and Arrow I had some issues with different behavior so I just left it on DataFrames for now.

krynju replied:
Added a Tables.jl compatible constructor + tests for CSV, Arrow, DataFrames

krynju commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,46 @@
+    partition_rows = (x, l, r) -> getindex(x, l:r)
+    df_create = (rows,i) -> DataFrames.DataFrame( partition_rows(rows, i, (i + thunksize - 1) % (n + 1)))
+    return DTable([Dagger.@spawn df_create(r,i) for i in 1:thunksize:n], thunksize)
+end
```
So currently if you run fetch() on a DTable (either just created or one that had filter ran on it) it will vcat the chunks back into a DataFrame

jpsamaroo commented:
I'm fine with moving this functionality out once it's ready to go; in the meantime we can keep it here for easy access to CI, and for testing in case we need/want to make scheduler changes as part of this work.

xiaodaigh commented:
Yeah. I think Dagger can exist without a table so I think the table functionality should be moved out as well. 

Now for practicality reason, perhaps it's ok to keep it here for CI reason etc.

vchuravy commented:
On alternative is to make it a subpackage, slightly annoying CI setup but makes joined development easier without bringing in unecessary dependencies or using Requires.

krynju commented:
@jpsamaroo I've been thinking about managing the chunks that contain the partitions in a scenario where the table would be bigger than the available memory. 
So at this stage this implementation doesn't care and loads all the partitions into memory, so naturally the process will die when it goes above the limit (at least when using threads, didn't check with processes)

So my idea for now is to first adjust the chunksize to memory size instead of rowcount to enable the rest of the functionality.
Then there's two scenarios:
1. Tables loaded using the Tables.jl interface - the table will need to be partitioned within the constructor according to the chunksize. There will have to be a mechanism of caching some chunks into disk cache and then loading them back on demand.
2. Tables that were initialized with a set of files - these will initially be partitioned according to the files provided and loaded into memory on demand. Then in case of any repartitioning (grouping, sorting etc.) the table will have to be treated as the table from point 1, so it will be partitioned and cached to the disk normally.

So both of these assume we would have some form of visibility on the memory available for a Processor to figure out when to cache the chunks and when not to load them into memory. I would probably declare/infer that information on the OSProc level, so even in a Distributed environment that memory availability could be handled. A different problem is where to cache the data - local env is easy, but in Distributed we should probably cache the table chunks at some commonly available location.

I'm wondering how much of this should be a part of the Dagger internals/scheduler and what should be a part of the DTable. I could imagine having a CachableChunk type for which you could define cache/restore functions and the scheduler figuring out when to call the cache and restore on these in order to make space in the physical memory and prepare the chunks for the Tasks to execute. With something like that the DTable wouldn't have a lot of work to do to provide out of core processing.

What's your take on this?

jpsamaroo commented:
I've thought about swap-to-disk a lot recently; some initial thoughts are here: https://github.com/JuliaParallel/Dagger.jl/issues/214#issuecomment-847926781, which sound quite similar to what you're suggesting.

> I'm wondering how much of this should be a part of the Dagger internals/scheduler and what should be a part of the DTable.

This is definitely a thing that the scheduler should know about and handle automatically. It's also relevant to the DArray, and generally to other kinds of operations that generate lots of data.

> I could imagine having a CachableChunk type for which you could define cache/restore functions and the scheduler figuring out when to call the cache and restore on these in order to make space in the physical memory and prepare the chunks for the Tasks to execute.

We already have the cache/restore in the form of checkpointing available to thunks (https://github.com/JuliaParallel/Dagger.jl/issues/183), although this is more of an automatic operation, whereas checkpointing is more manual (although that can change). Regardless, I don't think we need a new chunk type; the existing `Chunk` has a `handle` field that can be a `MemPool.FileRef`, or something else of our construction.

> With something like that the DTable wouldn't have a lot of work to do to provide out of core processing.

It's my hope that we can handle all of these details in the scheduler, so that any Dagger computation can benefit from swap-to-disk. I don't think exposing it to the DTable would be necessary; any knobs could be controlled via multiple dispatch on the chunked data type.

krynju commented:
So essentially there's three separate topics:
1. Adding the scheduler memory awareness and management 
2. Chunk with a handle type specialized for loading parts of the table stored initially as a set of files reachable by some Processors. That would come with a user defined loader, so they can use any input format as long as at the end it produces a Tables.jl compatible source ingestible by the DataFrame.
3. Chunk with a handle type specialized for storing table partitions in memory with specialized cache/restore functionality, so that the Scheduler can manage these according to the memory limits on used Processors.

So 2 and 3 is separate, because in my mind a chunk from point 2 becomes a chunk from point 3 after being successfully loaded once (and in case the partition gets dropped accidentaly it can be loaded again from that source).

jpsamaroo commented:
> Chunk with a handle type specialized for loading parts of the table stored initially as a set of files reachable by some Processors. That would come with a user defined loader, so they can use any input format as long as at the end it produces a Tables.jl compatible source ingestible by the DataFrame.

> Chunk with a handle type specialized for storing table partitions in memory with specialized cache/restore functionality, so that the Scheduler can manage these according to the memory limits on used Processors (other chunks would be just serialized/deserialized for memory management purposes, here we could optimize it for the partitions specifically).

I think we need that handle type (which can just be `MemPool.FileRef` potentially), and potentially also some way of expanding the `Processor` abstraction to reflect persistent storage. So the delineation would be:
- Chunk with handle of type `MemPool.DRef` (what's currently used to represent in-memory data), and processor that the data is associated with.
- Chunk with handle of type `MemPool.FileRef`, and "processor" representing the storage format of the data.

Here, the `DRef` and `FileRef` implement the means to retrieve stored data; doing `MemPool.poolget(::DRef)` returns the data stored in memory, while `MemPool.poolget(::FileRef)` should *ideally* return a representation of the data as stored on the file system (probably an open file descriptor on the host system). (Note that `MemPool.poolget` is called by `Dagger.move` by default). In the case of `FileRef`, it *should not* actually read the data into memory: that should be the job of `Dagger.move`, which will convert and serialize the data as needed. For a `FileRef`, doing a `Dagger.move` from some "storage processor" to a `ThreadProc` would imply reading the data into memory on that thread (really, just into the associated Julia process) in the most efficient manner possible. The combination of source processor, destination processor, and data type type should direct everything necessary to move data from file to memory, and back.

So what's missing here, in my mind, as an extra abstract type `Storage`, subtypes of which can be used as the `processor` field of `Chunk`, and represents that instead of having the data be associated with a computational device, it's associated with a storage device. It wouldn't be legal to execute code against a `Storage` type, so there'd be an extra step required to make a new `Chunk` with a real `Processor` (in other words, move it into memory).

> Adding the scheduler memory awareness and management

Given the above, the scheduler could be taught to move data from disk to memory when the data is necessary for a computation. It can also be taught that, when bringing in a new piece of data into memory in some way, if the memory requirement estimated for that data is greater than the memory limit of the processor (CPU RAM, GPU VRAM, etc.), then other chunks that're in-memory right now will need to first be moved to disk and cleared from memory.

An inability to perform this operation (say, because disk space ran out) would result in the scheduler setting an error state on that task of the DAG and anything that depends on it. In an interactive setting, the user could clean up more disk space *or* memory, and then re-run the computation and the scheduler should then retry from where it left off.

So what will be the format on disk for some arbitrary blob of data currently in-memory? This should probably be determined by a hypothetical `Dagger.default_storage_type(X)`, which would return a `Storage` subtype that can feasibly store `X`. Dagger would try to find a matching `Storage` instance, and if it succeeds, the scheduler could use that concrete instance to initiate a transfer into that storage location. The result of that transfer will be a `MemPool.FileRef`, which can later be used by the same `Storage` instance to retrieve the data from disk.


So how does all this work for the DTable? The DTable will need to auto-detect (based on the input table type or other mechanism) what `Storage` subtype is appropriate to represent the data is it is passed in, and then query Dagger to retrieve a matching storage instance. That can be used to construct any chunks associated with the input table, and those chunks can be safely passed as thunk arguments, and the scheduler will figure things out. Of course, the scheduler will then be free to temporarily send that data to disk in whatever format it wants (probably not CSV I would hope).


Thoughts?

krynju commented:
Makes sense to me with my current understanding of the internals

Although I have 3 questions:
1. Should the DTable storage stay as it is? Currently it's a `Vector{Dagger.EagerThunk}` and the thunk inside after creation is a simple `Dagger.@spawn (()->df)()`. If there's some operation applied the thunk will represent that operation instead. If I were to replace the storage with a `Vector{Chunk}` then I would have to unwrap the thunks with some table operations to retrieve the chunk and place it there, but I suppose that will break the execution graph if another operation were to be done on that retrieved chunk.
2. On creation (keeping in mind the new interface) I would have to explicitly create a chunk with .processor attribute set to the Storage instance the partition would be loaded from. What then? Should I wrap it in a simple thunk as well and keep it like that in the DTable vector? eg. `Dagger.@spawn collect(Dagger.tochunk(<partition>))`
3. If we ever wanted to give special treatment to the partitions caching/restore mechanism (maybe compression) how to specialize that? Implementing a `Dagger.move` operation for these chunks? How to figure out whether it is that partition chunk? New handle type?

shashi commented:
Is there any plans to support indexing? How would groupby and join work efficiently? I spoke to @quinnj  last year about possibly adding partitions and indexing helpers to Tables.. might to be a good time to restart that discussion.

jpsamaroo commented:
We might want to consider what's available with Tables and TableOperations now: https://youtu.be/Ryp0F0dHnLA

krynju commented:
I've seen that part of the interface and it will be very useful here both as a consumer of it and a source.

I've just added some small docs, fixed some consistency issues and put the DataFrames dependency in @requires. I think I can consider this initial scope ready as it is. 

Next todo's I think can be summarized by this list:

- [ ] Tables.jl (being a source) - figure out what parts of the interface and how to implement them in a way that makes sense.
- [ ] Tables.jl (consuming the interface) - add a constructor consuming the partitions interface
- [ ] Prepare a strategy on how to tackle repartitioning related operations: shuffle, sort, groupby, join; and assess their memory consumption.
- [ ] select and transform
- [ ] Implement the Storage API into the DTable when it's ready
- [ ] Figure out indexing
- [ ] Figure out potential schema usage/discovery and whether it can be any useful or just keep it unknown
- [ ] Add some convenience functions to inspect the contents of the DTable such as getindex, getcolumn, first, last, describe, schema information, length, partition information, memory placement across Processors, cached percentage etc.
- [ ] Consider some form of DResult structure for returning results of operations first applied per chunk, but not yet aggregated. Useful in map if you don't want to immediately get a full vcatted vector on the master processor (not so much in reduce).

quinnj commented:
This is exciting! Feel free to ping me on Tables.jl review stuff; happy to help. Or even to brainstorm the implementation, happy to chat on slack or discourse.

jpsamaroo commented on src/table/operations.jl:
````diff
@@ -0,0 +1,85 @@
+```
+"""
+function map(f, d::DTableRows)
+    chunk_wrap = chunk -> begin 
+        i = eachrow(chunk)
+        map(f, i)    
+    end
````
Although these really should be pulled out into dedicated functions, because otherwise we'll incur a lot of compile time (due to specializing the closure object on the captured `f`).

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,85 @@
+   
+    chunk_results = [Dagger.@spawn chunk_wrap(c) for c in d.dtable.chunks]
+    
+    # this is dumb, probably a better way to do this
+    # it gets the function used to aggregate x and y in a reduction function 
+    # eg. for: (x,y) -> x + y; this thing will return Base.+ 
+    get_aggregate_fun = (fun) -> begin
+        ast_f = Base.uncompressed_ast(methods(fun).ms[1])
+        return_line = ast_f.code[end]
+        if hasproperty(return_line, :val) 
+            ast_id_of_ret_val = return_line.val.id # julia 1.6+
+        else
+            ast_id_of_ret_val= return_line.args[1].id # julia 1.5
+        end
+        ast_agg_op = ast_f.code[ast_id_of_ret_val].args[1]
+        getfield(ast_agg_op.mod, ast_agg_op.name)
+    end
```
Yeah, this is no good. What's the reasoning for this?

krynju replied:
Well, if I assume the format of the user provided function to be 
`(acc, row) -> any_associative_fun(acc, any_user_transform(row))`
then I would want to figure out the `any_associative_fun` out of it to reduce the values obtained from reducing chunks.

I've been thinking about it all morning, but I think I've answered my question right as I was writing this answer.
I think instead of asking the user to provide a generic reduce function I'd rather ask them to provide `any_associative_fun` and `any_user_transform` separately. That would let me avoid this AST mess

I'll try that instead

krynju replied:
Replaced this monstrosity with an interface change to provide two separate functions

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,85 @@
+end
+
+"""
+    function filter(f, d::DTable)
```
```suggestion
    filter(f, d::DTable)
```

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,85 @@
+"""
+    function filter(f, d::DTable)
+
+Filter the `DTable` using `f`.
```
```suggestion
Filter `d` using `f`.
```

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,85 @@
+    function filter(f, d::DTable)
+
+Filter the `DTable` using `f`.
+Returns a filtered DTable that can be processed further.
```
```suggestion
Returns a filtered `DTable` that can be processed further.
```

jpsamaroo reviewed (changes requested):
Looking good, although we need to be careful about using closures, since they can add unacceptable compile time costs.

krynju commented on src/table/dresult.jl:
```diff
@@ -0,0 +1,17 @@
+struct DResult
```
Oops, that's a prototype file I made for one of the ideas from the todo list, it's not referenced anywhere though, so I'll remove it for now.
Basically it's supposed to represent a result of an operation that is first applied per chunk and then aggregated using some function. For example: map is first applied per chunk and then result gets aggregated using vcat when you call fetch on it.
It would be useful if you never wanted to collect the result into memory and just save it to the disk, but apart from map I didn't really find any useful scenarios for usage of this

krynju replied:
Removed for now, topic is in todo

krynju commented on src/table/operations.jl:
```diff
@@ -0,0 +1,85 @@
+end
+
+function map(f, d::DTableColumns)
+    throw("Not implemented")
```
I left it out intentionally for now. I don't ever do these kinds of operations, but I know it's a thing, so I just put a placeholder. Seems to me that implementation of this might get more complicated. I'll look into it at some point, but not in this PR

krynju commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,116 @@
+_retrieve(x::Dagger.Chunk) = collect(x)
+
+function getcolumn(d::DTable, s::Symbol)
+    _f = (x) -> Dagger.@spawn getindex(x, :, s)
```
I'll check, but at some point somewhere it didn't work for me, so I went straight for getindex

krynju replied:
yeah, the macro doesn't capture x as a chunk/thunk argument, so it calls getindex on a chunk

```
julia> Dagger.getcolumn(d, :d)
ERROR: MethodError: no method matching getindex(::Dagger.Chunk{DataFrames.DataFrame, MemPool.DRef, OSProc}, ::Colon, ::Symbol)

julia> @macroexpand Dagger.@spawn d[:, :a]
:(d[:, :a])

julia> @macroexpand Dagger.@spawn getindex(d, :, :a)
quote
    #= /home/krynju/.julia/dev/Dagger/src/thunk.jl:275 =#
    var"#6#args" = (d, :, :a)
    #= /home/krynju/.julia/dev/Dagger/src/thunk.jl:276 =#
    (Dagger.spawn)(getindex, var"#6#args"...; )
end
```

krynju commented on src/table/operations.jl:
```diff
@@ -0,0 +1,85 @@
+        map(f, i)    
+    end
+    result_vector = [Dagger.@spawn chunk_wrap(c) for c in d.dtable.chunks]
+    Dagger.@spawn ((r...) -> vcat((r)...))(result_vector...)
```
leftover mess from having Dagger.@spawn in there as well :smile:

krynju commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,116 @@
+    if !Tables.istable(table)
+        throw(ArgumentError("Provided input is not Tables.jl compatible."))
+    end
+    create_chunk = (rows) -> begin 
```
Extracted it into separate a separate function into a file with all dataframes dependencies

krynju commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,116 @@
+    chunks = Vector{Dagger.Chunk}()
+    sizehint!(chunks, length(files))
+
+    _load = file -> loader_function(file)
```
Fixed, leftover from having @spawn there

krynju commented on src/Dagger.jl:
```diff
@@ -65,6 +65,10 @@ function __init__()
     end
+    @require DataFrames="a93c6f00-e57d-5684-b7b6-d8193f3e46c0" begin
+         # Distributed table 
+         include("table/dtable.jl")
```
Resolved

krynju commented:
@jpsamaroo applied requested changes
CI failing due to some unrelated issue on master

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+const VTYPE = Vector{Union{Dagger.Chunk,Dagger.EagerThunk}}
+
+"""
+    struct DTable
+
+Structure representing the distributed table based on Dagger.
+
+The table is stored as a vector of `Chunk` structures which hold partitions of the table.
+That vector can also store `EagerThunk` structures when an operation that modifies
+the underlying partitions was applied to it (currently only `filter`).
+
+Underlying partitions should always be `DataFrame` structures.
```
```suggestion
    DTable

Structure representing the distributed table based on Dagger.

The table is stored as a vector of `Chunk` structures which hold partitions of the table.
That vector can also store `EagerThunk` structures when an operation that modifies
the underlying partitions was applied to it (currently only `filter`).
```

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+import Tables
+import TableOperations
+
+import Base:fetch
```
```suggestion
import Base: fetch
```

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+    chunks::VTYPE
+    tabletype
+    DTable(chunks::VTYPE, tabletype) = new(chunks, tabletype)
+    DTable(chunks::Vector{Dagger.EagerThunk}, args...; kwargs...) = new(VTYPE(chunks), args...; kwargs...)
+    DTable(chunks::Vector{Dagger.Chunk}, args...; kwargs...) = new(VTYPE(chunks), args...; kwargs...)
```
```suggestion
    DTable(chunks::Vector{Dagger.EagerThunk}, args...; kwargs...) = DTable(VTYPE(chunks), args...; kwargs...)
    DTable(chunks::Vector{Dagger.Chunk}, args...; kwargs...) = DTable(VTYPE(chunks), args...; kwargs...)
```

Internal ctors don't take kwargs, so we need to forward to external ctors instead. Maybe move these out of the struct body?

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+It assumes no initial partitioning of the table and uses the `chunksize`
+argument to partition the table (based on row count).
+"""
+function DTable(table, chunksize::Integer;tabletype=nothing)
```
```suggestion
function DTable(table, chunksize::Integer; tabletype=nothing)
```

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+    parts = Tables.partitions(TableOperations.makepartitions(table, chunksize))
+    chunks = Vector{Dagger.Chunk}()
+    sizehint!(chunks, length(parts))
+    _transform = Tables.materializer(table)
+    if tabletype !== nothing
+        _transform = Tables.materializer(tabletype())
+    end
```
```suggestion
    _transform = Tables.materializer(tabletype !== nothing ? tabletype() : table)
```

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+    if tabletype !== nothing
+        _transform = Tables.materializer(tabletype())
+    end
+    append!(chunks, Dagger.tochunk.(_transform.(parts)))
```
This isn't ideal to do directly, but it's fine for now. Let's remember to use `Dagger.@spawn` for this in a future PR.

krynju replied:
There would be two reasons to load it from the main process directly, so maybe there can be two separate constructors in the future for that.
1. Maybe only the main process has access to local directory/network location from which the data is being loaded
2. If there was ever a need to split files into partitions as well due to them being too big (without implementing the chunk vector as some kind of tree).

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+        _transform = Tables.materializer(tabletype())
+    end
+    append!(chunks, Dagger.tochunk.(_transform.(parts)))
+    return DTable(chunks, if tabletype === nothing typeof(table) else tabletype end)
```
```suggestion
    return DTable(chunks, tabletype === nothing ? typeof(table) : tabletype)
```

It's mostly just preference, but I prefer a ternary expression when it's all on one line and very simple.

jpsamaroo commented on src/table/dtable.jl:
````diff
@@ -0,0 +1,110 @@
+    DTable(Arrow.Table, ["a.arrow", "b.arrow"])
+```
+"""
+function DTable(loader_function, files::Vector{String};tabletype=nothing)
````
```suggestion
function DTable(loader_function, files::Vector{String}; tabletype=nothing)
```

jpsamaroo commented on src/table/dtable.jl:
````diff
@@ -0,0 +1,110 @@
+
+# Examples
+```jldoctest
+    DTable(CSV.File, ["a.csv", "b.csv"])
+    DTable(Arrow.Table, ["a.arrow", "b.arrow"])
````
Won't these files need to exist for the doctest? If so, maybe we can generate and write them out in the doctest, and then read them back in?

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+    sizehint!(chunks, length(files))
+    sizehint!(types, length(files))
+
+    _transform = Tables.materializer(loader_function(files[1]))
+    if tabletype !== nothing
+        _transform = Tables.materializer(tabletype())
+    end
```
```suggestion
    _transform = tabletype !== nothing ? Tables.materializer(tabletype()) : Tables.materializer(loader_function(files[1]))
```

jpsamaroo replied:
I also don't prefer the `files[1]` thing, but it's fine for now. In a future PR, we can do this lazily with `Dagger.@spawn` and figure it out in the task.

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+        push!(types, typeof(tpart))
+    end
+
+    @assert all(x-> x == types[1], types)
```
In the future, let's plan to remove this requirement.

krynju replied:
It's out for now, but some form of check in some operations will definitely be needed in case of some operations requiring non-generic functionality

jpsamaroo commented on src/table/dtable.jl:
```diff
@@ -0,0 +1,110 @@
+    jp = TableOperations.joinpartitions(Tables.partitioner(_retrieve, d.chunks))
+end
+
+function fetch(d::DTable, sink)
+    sink(fetch(d))
+end
```
```suggestion
fetch(d::DTable, sink) = sink(fetch(d))
```

jpsamaroo commented on src/table/operations.jl:
````diff
@@ -0,0 +1,60 @@
+```
+"""
+function map(f, d::DTableRows)
+    chunk_wrap = chunk -> map(f, eachrow(chunk))
````
This should take `f` and `chunk` as arguments. I've found that closures like these will get specialized on `f`, which can cause excessive compilation when playing around in the REPL.

jpsamaroo commented on src/table/operations.jl:
````diff
@@ -0,0 +1,60 @@
+```
+"""
+function reduce(f_reduce::Function, f_row::Function, d::DTableRows; init)
+    chunk_reduce = (agg, row) -> f_reduce(agg, f_row(row))
````
Should take `f_reduce` and `f_row` as arguments.

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,60 @@
+    chunk_reduce = (agg, row) -> f_reduce(agg, f_row(row))
+    chunk_wrap = (chunk, f, init) -> reduce(f, eachrow(chunk); init=init)
+    chunk_results = [Dagger.@spawn chunk_wrap(c, chunk_reduce, init) for c in d.dtable.chunks]
+    reduce_chunks = (a...) -> reduce(f_reduce, a; init)
+    Dagger.@spawn reduce_chunks(chunk_results...)
```
Should take `f_reduce` and `init` as arguments.

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,60 @@
+end
+
+"""
+    filter(f, d::DTable)
```
```suggestion
    filter(f, d::DTable) -> DTable
```

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,60 @@
+end
+
+"""
+    reduce(f_reduce::Function, f_row::Function, d::DTableRows; init)
```
```suggestion
    reduce(f_reduce::Function, f_row::Function, d::DTableRows; init) -> DTable
```

krynju replied:
Doesn't return a DTable if that's the meaning

jpsamaroo commented on src/table/operations.jl:
```diff
@@ -0,0 +1,60 @@
+import Base: filter, map, reduce
+
+"""
+    map(f, d::DTableRows)
```
```suggestion
    map(f, d::DTableRows) -> DTable
```

krynju replied:
Doesn't return a DTable if that's the meaning

jpsamaroo commented on test/table.jl:
```diff
@@ -0,0 +1,101 @@
+using Test
```
```suggestion
```

jpsamaroo commented on test/table.jl:
```diff
@@ -0,0 +1,101 @@
+        s = sample_df()
+        d = DTable(s, 100)
+        @test fetch(d) == s
+        end
```
```suggestion
    end
```

jpsamaroo commented on test/table.jl:
```diff
@@ -0,0 +1,101 @@
+        @test fetch(d) == DataFrame(t)
+    end
+
+        @testset "constructors - Tables.jl compatibility (CSV)" begin
```
```suggestion
    @testset "constructors - Tables.jl compatibility (CSV)" begin
```

jpsamaroo commented on test/table.jl:
```diff
@@ -0,0 +1,101 @@
+        @test fetch(dr4) == sr4
+    end
+end
+        
```
```suggestion
```

jpsamaroo reviewed (approved):
Great work! Just a few nitpicks and comments, and then I'd be happy to merge this! :smile:

jpsamaroo commented:
Thanks a ton! I'm looking forward to seeing this develop :smile:
